pymmcore_plus/core/_mmcore_plus.py

```diff
@@ -41,9 +43,18 @@ def __init__(self, mm_path=None, adapter_paths: ListOrTuple[str] = ()):
         self._canceled = False
         self._paused = False
 
+        # use weakref to avoid atexit keeping us from being
+        # garbage collected
+        self._weak_clean = weakref.WeakMethod(self.unloadAllDevices)

```

Comment:
i wonder whether [`weakref.finalize`](https://docs.python.org/3/library/weakref.html#weakref.finalize) might solve this for both the del and the exit case?

Reply:
otherwise i'm good with this PR.  Can you look into this briefly before we merge?

Reply:
> It is important to ensure that func, args and kwargs do not own any references to obj, either directly or indirectly, since otherwise obj will never be garbage collected. In particular, func should not be a bound method of obj.

Seems to imply that we'd still need a `WeakMethod`

Reply:
googling for examples perhaps I've misunderstood though

Reply:
> Seems to imply that we'd still need a WeakMethod

ah, it sure does.  fine with me to go with what you've got

Reply:
yeah I tried for a bit and couldn't get it working with finalize - seemed like it wasn't always being called for reasons I don't fully get.

Reply:
thanks for digging! LGTM